### PR TITLE
Add support for relationship-level links and meta attributes.

### PR DESCRIPTION
**Pull request by beauby**
Based on top of #1447.
Allows the following:
```ruby
has_many :posts do
  link :self do
    href '//example.com/link_author/relationships/posts'
    meta stuff: 'value'
  end
  link :related do
    href '//example.com/link_author/posts'
    meta count: object.posts.count
  end
  include_data false
end
```

### CODE REVIEW

**Comment by beauby** [comment]
Travis is failing because of the rubocop config (https://github.com/rails-api/active_model_serializers/blob/master/.rubocop.yml#L40 – we should allow `with_first_parameter`).

File: lib/active_model/serializer/adapter/json_api.rb [hunk 1]
```diff
@@ -80,61 +88,45 @@ def fragment_cache(cached_hash, non_cached_hash)
...
-          end
+        def resource_objects_for(serializers)
+          @primary = []
+          @included = []
+          @resource_identifiers = Set.new
+          serializers.each { |serializer| process_resource(serializer, true) }
+          serializers.each { |serializer| process_relationships(serializer, @include_tree) }
 
-          hash
+          [@primary, @included]
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api.rb]: any reason for the instance variables here?

File: lib/active_model/serializer/adapter/json_api/association.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+            @options = options
+            @data = data_for(serializer, options)
+            @links = links
+                     .map { |key, value| { key => Link.new(parent_serializer, value).as_json } }
+                     .reduce({}, :merge)
+            @meta = meta.respond_to?(:call) ? parent_serializer.instance_eval(&meta) : meta
+          end
+
+          def as_json
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/association.rb]: :+1:

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/association.rb]: is this an APIObject?

File: lib/active_model/serializer/adapter/json_api/resource_identifier.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        class ResourceIdentifier
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/resource_identifier.rb]: ApiObject?

File: lib/active_model/serializer/association.rb [hunk 4]
```diff
@@ -9,7 +9,7 @@ class Serializer
     # @example
     #  Association.new(:comments, CommentSummarySerializer)
     #
-    Association = Struct.new(:name, :serializer, :options) do
+    Association = Struct.new(:name, :serializer, :options, :links, :meta) do
```
Comment (bf4) [inline lib/active_model/serializer/association.rb]: JSON API concerns creeping into the serializer.. hmm

File: test/adapter/json_api/links_test.rb [hunk 5]
```diff
@@ -17,11 +17,23 @@ class LinkAuthorSerializer < ActiveModel::Serializer
             link :yet_another do
               "//example.com/resource/#{object.id}"
             end
+
+            has_many :posts do
+              link :self do
+                href '//example.com/link_author/relationships/posts'
```
Comment (bf4) [inline test/adapter/json_api/links_test.rb]: shouldn't these be auto-generated in most cases?

**Comment by bf4** [comment]
Mostly looks good, though, since both there are boundary changes along with refactoring and new behavior, I'll want to come back and look at what has 'changed'

**Comment by groyoh** [comment]
Any update on this PR? Is there something needed to merge this PR? I could work on it if needed.

File: lib/active_model/serializer/adapter/json_api/resource_identifier.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          def type_for(serializer)
+            return serializer._type if serializer._type
+            if ActiveModelSerializers.config.jsonapi_resource_type == :singular
+              serializer.object.class.model_name.singular
+            else
+              serializer.object.class.model_name.plural
+            end
+          end
+
+          def id_for(serializer)
```
Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/resource_identifier.rb]: should be `serializer.read_attribute_for_serialization(:id)`

**Comment by bf4** [comment]
@groyoh that's be great.   Would you want to make a PR into @beauby 's branch?  I can review and merge that.  I think he's not around right now.

File: lib/active_model/serializer/adapter/json_api/resource_identifier.rb (hunk 6 above)
Comment (groyoh) [inline lib/active_model/serializer/adapter/json_api/resource_identifier.rb]: To be on the same page here, that would change the current behaviour, right? (see https://github.com/beauby/active_model_serializers/blob/master/lib/active_model/serializer/adapter/json_api.rb#L125)

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/resource_identifier.rb]: nope, I did indeedy mean to call it on the serializer, but @groyoh thanks for pointing out my error that this was just a direct port of the method

Comment (groyoh) [inline lib/active_model/serializer/adapter/json_api/resource_identifier.rb]: Oh you're right, I didn't realized that `read_attribute_for_serialization` was now defined on the serializer.

File: lib/active_model/serializer/adapter/json_api/association.rb (hunk 2 above)
Comment (groyoh) [inline lib/active_model/serializer/adapter/json_api/association.rb]: @bf4 should this be namespaced under `module ApiObjects`?

Comment (bf4) [inline lib/active_model/serializer/adapter/json_api/association.rb]: if we follow that pattern, yes. It was always about how we feel about it making sense https://github.com/rails-api/active_model_serializers/pull/1004#discussion_r46783510 and https://github.com/rails-api/active_model_serializers/pull/1482

**Comment by bf4** [comment]
Merged. Added Needs follow up tag.

**Comment by janvarljen** [comment]
is there a way to use Rails URL helpers in links?

like this:
```ruby
has_many :posts do
  link :related do
    href api_author_posts_url(author: object)
  end
  include_data false
end
```

**Comment by sheldonnbbaker** [comment]
@janvarljen I did the following hack to get what you're after.

```ruby
module ActiveModel
  class Serializer
    module Adapter
      class JsonApi
        class Link
          def routes
            Rails.application.routes.url_helpers
          end
        end
      end
    end
  end
end
```

```ruby
has_many :posts do
  link :related do
    href routes.api_author_posts_url(author: object)
  end
  include_data false
end
```

**Comment by groyoh** [comment]
@sheldonnbbaker Nice hack but I'd recommend:

```ruby
UrlHelpers = Rails.application.routes.url_helpers
```
```ruby
has_many :posts do
  link :related do
    href UrlHelpers.api_author_posts_url(author: object)
  end
  include_data false
end
```

For two reasons:

1. No monkey patch
2. Prevents you from instantiating a module every time your call the routes (https://github.com/rails/rails/blob/v4.2.5/actionpack/lib/action_dispatch/routing/route_set.rb#L490)

@janvarljen I don't think there is a cleaner way at the moment.

**Comment by bf4** [comment]
@sheldonnbbaker @janvarljen @remear @groyoh this url_helpers issue is https://github.com/rails-api/active_model_serializers/issues/1269 just needs a PR.. good stuff above

**Comment by janvarljen** [comment]
@sheldonnbbaker @groyoh thx for the ideas. I'm doing the same thing atm but wanted to discuss a more elegant solution. 

One more thing I think we're missing in these solutions is url_options context like :host, :protocol etc... we need for url_helpers to work as expected.

They way I do it:

in api_controller
```ruby
serialization_scope :scope

def scope
  {
    url_options: url_options
  }
end
```

in base_serializer
```ruby
include Rails.application.routes.url_helpers

def url_options
  scope[:url_options]
end
```

then I can use url helpers in serializer methods but not inside link block because we don't have access to the serializer methods

**Comment by remear** [comment]
I'm working on adding the url_helpers to 0.10 so you won't need a workarounds like these. If all goes well, that means you should have the ability to do things like:

```ruby
link :self { user_url(object) }
link :self { url_for([object, :users, @user]) }
link :self { "http://example.com/posts/#{object.id}" }
link :self, "http://example.com/user"
```

**Comment by janvarljen** [comment]
@remear that would be great. Ping me if you need any help

**Comment by bf4** [comment]
ref: https://github.com/rails-api/active_model_serializers/issues/1269

**Comment by bf4** [comment]
Maybe a 'wip' pr?

**Comment by remear** [comment]
@janvarljen, @sheldonnbbaker Feedback is welcome on #1550 that aims to add url_helpers for links.

**Comment by rjsteen** [comment]
Does anyone know when this functionality will be available in active_model_serializers gem?

**Comment by NullVoxPopuli** [comment]
you can install from our master branch until we cut another release -- would that work?

**Comment by rjsteen** [comment]
I am not seeing these changes on `active_model_serializers:master`

**Comment by remear** [comment]
This PR and #1550, which adds Rails url_helpers, have been merged into the master branch. They are not yet available in a version available from RubyGems. You'll need to have your application use the master branch of this repository until the next release. If you're having other troubles, I suggest opening a new issue with more details.

**Comment by rjsteen** [comment]
:+1:

**Pull request merged**